### PR TITLE
Call refresh() if the first response is 401

### PR DESCRIPTION
**Pull request by mateuszwojtczak**
Related issue: https://github.com/dart-lang/oauth2/issues/119

### CODE REVIEW

File: packages/login_client/lib/src/login_client.dart [hunk 1]
```diff
@@ -108,6 +108,7 @@ class LoginClient extends http.BaseClient {
   /// This method will log the [LoginClient] out on the authorization failure.
   Future<void> logIn(AuthorizationStrategy strategy) async {
     try {
+      _oAuthClient?.close();
```
Comment (shilangyu) [inline packages/login_client/lib/src/login_client.dart:111]: Reverting #25?

Comment (mateuszwojtczak) [reply packages/login_client/lib/src/login_client.dart:111]: good catch!

File: packages/login_client/test/login_client_test.dart [hunk 2]
```diff
@@ -159,11 +159,62 @@ void main() {
...
+
+    test(
+      'send() refreshes credentials on 401 response',
+      () async {
+        final request = Request('POST', Uri.parse('www.example.com'));
+        final mockOauthClient = MockClient();
+
+        when(mockOauthClient.send(any))
+            .thenAnswer((_) async => StreamedResponse(
+                  Stream.fromIterable([]),
```
Comment (Albert221) [inline packages/login_client/test/login_client_test.dart:171]: `Stream.empty()`

Comment (mateuszwojtczak) [reply packages/login_client/test/login_client_test.dart]: good catch

File: packages/login_client/lib/src/login_client.dart [hunk 3]
```diff
@@ -196,4 +209,28 @@ class LoginClient extends http.BaseClient {
   // ignore: use_setters_to_change_properties
   @visibleForTesting
   void setAuthorizedClient(oauth2.Client client) => _oAuthClient = client;
+
+  http.BaseRequest _copyRequest(http.BaseRequest request) {
```
Comment (Albert221) [inline packages/login_client/lib/src/login_client.dart:213]: This method could be `static`

File: packages/login_client/lib/src/login_client.dart [hunk 4]
```diff
@@ -196,4 +209,28 @@ class LoginClient extends http.BaseClient {
...
+    if (request is http.Request) {
+      newRequest = http.Request(request.method, request.url)
+        ..encoding = request.encoding
+        ..bodyBytes = request.bodyBytes;
+    } else if (request is http.MultipartRequest) {
+      newRequest = http.MultipartRequest(request.method, request.url)
+        ..fields.addAll(request.fields)
+        ..files.addAll(request.files);
+    } else {
+      throw Exception('Request type is unknown, cannot copy');
```
Comment (Albert221) [inline packages/login_client/lib/src/login_client.dart:225]: It's not unknown. It's unsupported :)

**Review by Albert221 — APPROVED** [review]

**Pull request merged**